src/rust/lib/graph-editor/src/component/node/port.rs

```diff
@@ -0,0 +1,468 @@
+            // NOTE: `is_inwards` should only be 0.0 or 1.0.
+            new_port(height,width,inner_radius,is_inwards)
+        }
+    }
+
+    impl Shape{
+        /// Set the shape parameters derived from the `Specification`.
+        pub fn update_from_spec(&self,spec:&Specification){
+            self.height.set(spec.height);
+            self.inner_radius.set(spec.inner_radius);
+            self.width.set(spec.width.value.to_radians());
+            match &spec.direction{
+                Direction::In  => self.is_inwards.set(1.0),
+                Direction::Out => self.is_inwards.set(0.0),
+            };
+        }
+    }
+}
+
+
+
+// ============
+// === Port ===
+// ============
+
+/// Shape view for input port.
+#[derive(Debug,Default,Clone,Copy)]
+pub struct InputPortView {}
+impl ShapeViewDefinition for InputPortView {
+    type Shape = shape::Shape;
+    fn new(shape:&Self::Shape, _scene:&Scene,_shape_registry:&ShapeRegistry) -> Self {
+        let spec  = Specification::default();
+        shape.update_from_spec(&spec);
+        shape.is_inwards.set(1.0);
+
+        let bbox = Vector2::new(2.0 * shape.height.get(), 2.0 * shape.height.get());
+        shape.sprite.size().set(bbox);
+
+        Self {}
+    }
+}
+
+/// Shape view for output port.
+#[derive(Debug,Default,Clone,Copy)]
+pub struct OutputPortView {}
+impl ShapeViewDefinition for OutputPortView {
+    type Shape = shape::Shape;
+    fn new(shape:&Self::Shape, _scene:&Scene,_shape_registry:&ShapeRegistry) -> Self {
+        let spec  = Specification::default();
+        shape.update_from_spec(&spec);
+        shape.is_inwards.set(0.0);
+
+        let bbox = Vector2::new(2.0 * shape.height.get(), 2.0 * shape.height.get());
+        shape.sprite.size().set(bbox);
+
+        Self {}
+    }
+}
+
+/// Helper trait that describes a `ShapeViewDefinition` with a port shape.
+pub trait PortShapeViewDefinition = ShapeViewDefinition<Shape=shape::Shape>;
+
+/// Port definition.
+///
+/// A port must always be instantiated as an `InputPort` or an `OutputPort`. This determines its
+/// actual shape.
+///
+/// Example
+/// -------
+/// ```
+/// use graph_editor::component::node::port::InputPort;
+/// use graph_editor::component::node::port::OutputPort;
+///
+/// let input_port  = InputPort::default();
+/// let output_port = OutputPort::default();
+/// ```
+#[derive(Debug,CloneRef)]
+#[allow(missing_docs)]
+pub struct Port<T:PortShapeViewDefinition> {
+    pub data : Rc<PortData<T>>
+}
+
+impl<T:PortShapeViewDefinition> Clone for Port<T>{
+    fn clone(&self) -> Self {
+        let data = Rc::clone(&self.data);
+        Self {data}
+    }
+}
+
+/// Type that represents an input port.
+pub type InputPort = Port<InputPortView>;
+
+/// Type that represents an output port.
+pub type OutputPort = Port<OutputPortView>;
+
+/// Internal data of `Port.
+#[derive(Debug,Clone)]
+#[allow(missing_docs)]
+pub struct PortData<T:PortShapeViewDefinition> {
+        spec : RefCell<Specification>,
```

Comment:
IMO this `RefCell` is too broad. What I would propose is to remove the `Specification` and include all of its fields here. Many of the fields are `Copy` so they can be converted into `Cell`. Some fields are not `Copy`, but I think you would be fine (with the sea performance) to use `CloneCell` (from prelude) for them instead - please confirm if Im right.